### PR TITLE
Print error messages in case of parsing errors instead of failing an …

### PR DESCRIPTION
**Pull request by FrancescoPesce**
…assertion

### CODE REVIEW

File: lib/Codegen/Lowering/ForEquationLowerer.cpp [hunk 1]
```diff
@@ -67,8 +68,10 @@ namespace marco::codegen::lowering
       builder().setInsertionPointToStart(forEquationOp.bodyBlock());
 
       // Add the induction variable to the symbol table
+      const llvm::StringRef name = forIndex->getName();
+      getDeclaredVariables().insert(std::string(name));
```
Comment (mscuttari) [inline lib/Codegen/Lowering/ForEquationLowerer.cpp:72]: When is the variable going to be removed from the set?

File: lib/Codegen/Lowering/IdentifierError.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+#include "marco/Codegen/Lowering/Distance/EditDistance.h"
+#include "marco/Codegen/Lowering/Distance/SentenceDistanceCalculator.h"
+
+namespace marco::codegen::lowering
+{
+  IdentifierError::IdentifierError(const IdentifierType &identifierType, std::string identifierName, 
+        const std::set<std::string> &declaredIdentifiers, unsigned int line, unsigned int column): 
+        identifierType(identifierType), actualName(identifierName), predictedName(""),
+        line(line), column(column) {
+    unsigned int lowestDistance = UINT32_MAX;
```
Comment (mscuttari) [inline lib/Codegen/Lowering/IdentifierError.cpp:11]: "Lowest" and `UINT32_MAX` don't seem to be very coherent.

File: lib/Codegen/Lowering/IdentifierError.cpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    // Choose which built in identifiers to check.
+    const std::set<std::string> emptySet = {};
```
Comment (mscuttari) [inline lib/Codegen/Lowering/IdentifierError.cpp:14]: This is an ugly hack. Call appropriate functions passing the needed set or iterators.

File: lib/Codegen/Lowering/ClassLowerer.cpp [hunk 4]
```diff
@@ -48,7 +48,8 @@ namespace marco::codegen::lowering
     }
 
     if (auto function = cls.dyn_cast<ast::PartialDerFunction>()) {
-      return declareVariables(*function);
+      declareVariables(*function);
```
Comment (mscuttari) [inline lib/Codegen/Lowering/ClassLowerer.cpp:51]: Why is the result ignored?

File: lib/Codegen/Lowering/Distance/EditDistance.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        distances[index] = std::min(std::min(deletion_result, insertion_result), 
+                                    std::min(substitution_result, transposition_result));
+      }
+    }
+
+    return distances[(n+1)*(m+1)-1];
+  }
+
+  const unsigned int EditDistance::largest_substitution_cost = 21;
+  // Values obtained with a python script, using a measure of keyboard distance and
```
Comment (mscuttari) [inline lib/Codegen/Lowering/Distance/EditDistance.cpp:69]: This needs to be explained way better. Which Python script? What does it do? Why 2 as penalty?

File: lib/Codegen/Lowering/Distance/CMakeLists.txt
Comment (mscuttari) [inline lib/Codegen/Lowering/Distance/CMakeLists.txt:1]: How is failure managed? I don't see any conditional compilation depending on the satisfied dependencies and / or the outcome of the Python script.

File: lib/Codegen/Lowering/CMakeLists.txt [hunk 6]
```diff
@@ -48,4 +49,7 @@ marco_add_library(Lowering
     marco::ast
     marco::modeling
     marco::variableFilter
+    marco::Distance
```
Comment (mscuttari) [inline lib/Codegen/Lowering/CMakeLists.txt:52]: What if the wordnet stuff could not be built?

File: lib/Codegen/Lowering/CMakeLists.txt [hunk 7]
```diff
@@ -48,4 +49,7 @@ marco_add_library(Lowering
...
     ${MLIR_LIBS})
+
+add_subdirectory(Distance)
```
Comment (mscuttari) [inline lib/Codegen/Lowering/CMakeLists.txt:55]: This should be before the declaration of libraries depending on it.

File: lib/Codegen/Lowering/ComponentReferenceLowerer.cpp [hunk 8]
```diff
@@ -22,8 +22,22 @@ namespace marco::codegen::lowering
     const ast::ComponentReferenceEntry* firstEntry =
         componentReference.getElement(0);
 
-    Reference result = lookupVariable(firstEntry->getName());
-    result = lowerSubscripts(result, *firstEntry);
+    std::optional<Reference> optionalResult = lookupVariable(firstEntry->getName());
+    if (!optionalResult) {
+      std::set<std::string> declaredVars = {};
```
Comment (mscuttari) [inline lib/Codegen/Lowering/ComponentReferenceLowerer.cpp:27]: Why this local variable? Didn't you add the set to the class?

**Review by mscuttari — CHANGES_REQUESTED** [review]

File: include/public/marco/Codegen/Lowering/Distance/DatabaseReader.h [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace marco::codegen::lowering
+{
+    // We use the synset's hypernym.csv byte offset as the synset ID.
+    // Since synsets could also be defined in different ways, and one
+    // future implementation might use a different ID system, we use
+    // a typedef to make it easier to change the synset ID type.
+    using Synset = int;
+
+    class DatabaseReader
```
[the remark this answers is not in the document]
Comment (mscuttari) [reply include/public/marco/Codegen/Lowering/Distance/DatabaseReader.h:13]: I suggest having a look at the singleton pattern.

File: lib/Codegen/Lowering/Distance/CMakeLists.txt
Comment (michele-miotti-uni) [reply lib/Codegen/Lowering/Distance/CMakeLists.txt:1]: The outcome of the python script is not taken into consideration, except for a simple warning message. The change in the program's behavior depends on the content of the wordnet folder, which is left empty in case of failure. As soon as the program detects that the folder is empty at runtime, the semantic distance functionality is implicitly ignored.

File: lib/Codegen/Lowering/Distance/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+# As opposed to CMAKE_SOURCE_DIR, CMAKE_CURRENT_SOURCE_DIR is
+# the directory of the current CMakeLists.txt file. In this case,
+# it is the directory of this file, which is exactly where the
+# Python script is located.
+set(PYTHON_SCRIPT "${CMAKE_CURRENT_SOURCE_DIR}/install-db.py")
+
+# Custom command to run the Python script
+add_custom_command(
+    OUTPUT "${TARGET_DIR}/.timestamp"  # We use a timestamp file to signify that the script has run.
```
[the remark this answers is not in the document]
Comment (michele-miotti-uni) [reply lib/Codegen/Lowering/Distance/CMakeLists.txt:38]: ".timestamp" is a hidden empty file that signifies that the python "portion" of the process has already been executed, so there's no need to run it again. This is because the python script downloads a relatively large file, so needless repeated execution would be detrimental.

File: lib/Codegen/Lowering/Distance/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,58 @@
+set(SOURCES
+    DatabaseReader.cpp
+    EditDistance.cpp
+    SentenceDistanceCalculator.cpp
+    WordDistanceCalculator.cpp)
+
+marco_add_library(Distance
+    ${SOURCES}
+
+    LINK_LIBS PUBLIC
```
Comment (mscuttari) [inline lib/Codegen/Lowering/Distance/CMakeLists.txt:10]: Are these libraries needed?

File: lib/Codegen/Lowering/Distance/CMakeLists.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (mscuttari) [reply lib/Codegen/Lowering/Distance/CMakeLists.txt:36]: Can't the python script just search for previous results?

**Review by mscuttari — CHANGES_REQUESTED** [review]

File: lib/Codegen/Lowering/AlgorithmLowerer.cpp [hunk 12]
```diff
@@ -28,9 +28,13 @@ namespace marco::codegen::lowering
     builder().setInsertionPointToStart(algorithmBody);
 
     for (size_t i = 0, e = algorithm.size(); i < e; ++i) {
-      lower(*algorithm[i]);
+      const bool outcome = lower(*algorithm[i]);
```
[the remark this answers is not in the document]
Comment (FrancescoPesce) [reply lib/Codegen/Lowering/AlgorithmLowerer.cpp:31]: In a previous C++ course, we where told to always put const whenever possible. In which cases should we use it for this project? Only for class members and function parameters?

File: lib/Codegen/Lowering/StatementLowerer.cpp [hunk 13]
```diff
@@ -11,14 +11,15 @@ namespace marco::codegen::lowering
...
-  void StatementLowerer::lower(const ast::Statement& statement)
+  __attribute__((warn_unused_result)) bool StatementLowerer::lower(const ast::Statement& statement)
   {
     if (auto casted = statement.dyn_cast<ast::AssignmentStatement>()) {
       return lower(*casted);
     }
 
     if (auto casted = statement.dyn_cast<ast::BreakStatement>()) {
-      return lower(*casted);
+      lower(*casted);
```
[the remark this answers is not in the document]
Comment (FrancescoPesce) [reply lib/Codegen/Lowering/StatementLowerer.cpp:21]: In line 21, lower(*casted) returns void, since *casted is a BreakStatement. The outer function returns a bool, which is true if there was no error, and false otherwise. Since lower(*casted) returns void, we cannot simply return its result (since there is no result). Since no error can occur, true is returned.

[the remark this answers is not in the document]
Comment (mscuttari) [reply lib/Codegen/Lowering/StatementLowerer.cpp:21]: The previous return statement was used to future-proof the code with respect to API changes like the ones you are making. The lower method for BreakStatement should indeed be changed to have a status result, making it coherent with all the others.

File: include/public/marco/Codegen/Lowering/LoweringContext.h [hunk 14]
```diff
@@ -32,10 +34,14 @@ namespace marco::codegen::lowering
 
       mlir::OpBuilder& getBuilder();
 
+      mlir::DiagnosticEngine& getDiagEngine();
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/LoweringContext.h:37]: There should be no need for accessing the diagnostic engine at all. Errors can be emitted using the MLIR context.

File: lib/Codegen/Lowering/AlgorithmLowerer.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (mscuttari) [reply lib/Codegen/Lowering/AlgorithmLowerer.cpp:31]: Yes, use it only for class members and function parameters (and reference results, when they shall not be modified by the caller).

File: lib/Codegen/Lowering/CallLowerer.cpp [hunk 15]
```diff
@@ -123,17 +143,21 @@ namespace marco::codegen::lowering
         return nullptr;
       }
 
-      result = getSymbolTable().lookupSymbolIn(
```
Comment (mscuttari) [inline lib/Codegen/Lowering/CallLowerer.cpp:126]: Why this change?

File: lib/Codegen/Lowering/ExpressionLowerer.cpp [hunk 16]
```diff
@@ -11,7 +11,7 @@ namespace marco::codegen::lowering
   {
   }
 
-  Results ExpressionLowerer::lower(const ast::Expression& expression)
+  std::optional<Results> ExpressionLowerer::lower(const ast::Expression& expression)
```
[the remark this answers is not in the document]
Comment (FrancescoPesce) [reply lib/Codegen/Lowering/ExpressionLowerer.cpp:14]: I did not understand the comment. Is the unfinished work in this file? This is simply casting the expression to different classes and leaving the error handling to the respective Lowerers.

File: lib/Codegen/Lowering/Distance/DatabaseReader.cpp [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+#include <vector>
+
+namespace marco::codegen::lowering
+{
+    DatabaseReader* DatabaseReader::instance = NULL;
+    DatabaseReader* DatabaseReader::getInstance()
+    {
+        if (instance == NULL)
+        {
+            instance = new DatabaseReader();
```
Comment (mscuttari) [inline lib/Codegen/Lowering/Distance/DatabaseReader.cpp:14]: Do not manually allocate memory.

File: lib/Codegen/Lowering/ForEquationLowerer.cpp (hunk 1 above)
Comment (FrancescoPesce) [reply lib/Codegen/Lowering/ForEquationLowerer.cpp:71]: The variable is never removed from the set. "declaredVariables" contains every variable identifier that has ever been inserted into "variablesSymbolTable". When an error with a variable identifier occurs, there is a function to check, for each element of the set, whether it is visible from the location that generated the error or not.

File: lib/Codegen/Lowering/WhileStatementLowerer.cpp [hunk 18]
```diff
@@ -34,9 +34,13 @@ namespace marco::codegen::lowering
 
       const auto* condition = statement.getCondition();
 
+      auto optionalLoweredCondition = lower(*condition);
```
Comment (mscuttari) [inline lib/Codegen/Lowering/WhileStatementLowerer.cpp:37]: Remove `optional` from the name. Just an unneeded complication.

**Review by mscuttari — CHANGES_REQUESTED** [review]

File: lib/Codegen/Lowering/ForEquationLowerer.cpp (hunk 1 above)
Comment (mscuttari) [reply lib/Codegen/Lowering/ForEquationLowerer.cpp:71]: So, in general, the content of set is not coherent with the content of the symbol table. Let's make it coherent, as in future the set may be used for other reasons and the current state would lead to wrong behaviours.

File: lib/Codegen/Lowering/ExpressionLowerer.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (mscuttari) [reply lib/Codegen/Lowering/ExpressionLowerer.cpp:14]: My bad, I overlooked the kind of error, that would indeed be a compiler construction problem and not runtime related one.

File: lib/Codegen/Lowering/Distance/EditDistance.cpp (hunk 5 above)
Comment (FrancescoPesce) [reply lib/Codegen/Lowering/Distance/EditDistance.cpp:69]: The python script I am referring to is not in the repository. It is a simple script in which I wrote a simplified representation of a keyboard with the classic QWERTY layout, using a misclassification cost proportional to the physical distance between keys. The value of 2 has no mathematical meaning, it was simply found to work well experimentally. For a heuristic reason, 2 is also the cost for a mismatch for keys that adjacent in the same row. I will improve the comment.

File: include/public/marco/Codegen/Lowering/Distance/EditDistance.h [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+#define MARCO_CODEGEN_LOWERING_DISTANCE_EDITDISTANCE_H
+
+#include <array>
+#include <cassert>
+#include <optional>
+#include <string>
+
+namespace marco::codegen::lowering
+{
+  class EditDistance
```
[the remark this answers is not in the document]
Comment (FrancescoPesce) [reply include/public/marco/Codegen/Lowering/Distance/EditDistance.h:11]: It is a collection of functions, and parameters that are relative to the concept of "edit distance", so they were all put together into a single class. Should I remove the class and create an EditDistance namespace instead?

[the remark this answers is not in the document]
Comment (mscuttari) [reply include/public/marco/Codegen/Lowering/Distance/EditDistance.h:11]: Classes are made to be instantiated, otherwise they serve no purpose. Remove the class and let the methods live directly inside the namespace (`marco::codegen::lowering` is enough).

File: lib/Codegen/Lowering/ForEquationLowerer.cpp (hunk 1 above)
Comment (FrancescoPesce) [reply lib/Codegen/Lowering/ForEquationLowerer.cpp:71]: If I understood correctly, variablesSymbolTable has all of its elements removed when the destructor of LookupScopeGuard is invoked, and this is the only place where elements are removed from it. Therefore, I could simply modify the same destructor to remove all elements from declaredVariables as well. Is this correct?

**Review by mscuttari — CHANGES_REQUESTED** [review]

File: lib/Codegen/Lowering/ForEquationLowerer.cpp (hunk 1 above)
Comment (FrancescoPesce) [reply lib/Codegen/Lowering/ForEquationLowerer.cpp]: I wasn't referring to changing the `VariablesSymbolTable` destructor, but rather the `LookupScopeGuard` destructor. 
Since there is currently no way to iterate on the variables declared in a scope, this change would require storing the scope information in `declaredVariables` as well, basically reimplementing a `llvm::ScopedHashTable`.
For example, I could change `declaredVariables` into a map with scopes as keys, and `std::set<std::string>` as values, then edit the `LookupScopeGuard` destructor to remove all entries that have a scope in `lookupScopes` as keys. Is this what I should do?

Comment (mscuttari) [reply lib/Codegen/Lowering/ForEquationLowerer.cpp]: Declared variables and lookup scopes are somewhat close to each other but I would keep them separate for now. You can just make the current `VariablesSymbolTable` a new class wrapping the current `llvm::ScopedHashTable` and the `set` you need, and then implement a scope class along the same lines. This way you don't have to reimplement anything but a proxy.

Comment (FrancescoPesce) [reply lib/Codegen/Lowering/ForEquationLowerer.cpp]: I implemented the classes you suggested. 
However, the attribute `variablesSet` of `VariablesSymbolTable` still contains all variables, with no removals. The reason is that I realized that this information is needed elsewhere, since the same set is used in cases such as the one in `ParserError/fields/field_scope.mo`, to find the identifiers `fild` and `field`. In fact, while I can iterate on nested `mlir::SymbolTable`s, I cannot actually iterate on the variables declared in a single `mlir::SymbolTable`.
I believe the new code is still cleaner than the old one, as the only way to accidentally access variables that are not in the `llvm::ScopedHashTable` is using the new `getVariables` method of the `VariablesSymbolTable` with the input `onlyVisible = false`, and the set is now hidden inside the class.

File: include/public/marco/Codegen/Lowering/CallLowerer.h [hunk 20]
```diff
@@ -80,57 +80,64 @@ namespace marco::codegen::lowering
...
-      Results tan(const ast::Call& call);
-      Results tanh(const ast::Call& call);
-      Results transpose(const ast::Call& call);
-      Results zeros(const ast::Call& call);
-
-      Results reduction(const ast::Call& call, llvm::StringRef action);
+      /// Emit an error due to a wrong number of arguments being provided to a function.
+      void emitErrorNumArguments(llvm::StringRef function, const marco::SourceRange& location, 
+                                 unsigned int actualNum, unsigned int expectedNum);
+      void emitErrorNumArgumentsRange(llvm::StringRef function, const marco::SourceRange& location,
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/CallLowerer.h:86]: Whats the difference with respect to the other one?

File: include/public/marco/Codegen/Lowering/Distance/DatabaseReader.h [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/Distance/DatabaseReader.h:17]: What does this class do?

File: include/public/marco/Codegen/Lowering/Distance/EditDistanceParameters.h [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+    assert(character >= 'a' && character <= 'z');
+    return character - 'a' + 37;
+  }
+
+  // Costs for different edit operations.
+  constexpr unsigned int editDistanceBaseInsertionCost = 2;
+  constexpr unsigned int editDistanceDeletionCost = 2;
+  constexpr unsigned int editDistanceTranspositionCost = 3;
+  const unsigned int editDistanceLargestSubstitutionCost = 21;
+  const std::array<std::array<unsigned int, 63>, 63> editDistanceSubsitutionCosts =
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/Distance/EditDistanceParameters.h:34]: Where does this come from? I know the short answer, which is not sufficient. Other people don't even know that.

File: include/public/marco/Codegen/Lowering/Distance/EditDistanceParameters.h [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+    }
+    assert(character >= '_');
+    if (character == '_') {
+      return 36;
+    }
+    assert(character >= 'a' && character <= 'z');
+    return character - 'a' + 37;
+  }
+
+  // Costs for different edit operations.
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/Distance/EditDistanceParameters.h:29]: Explain better what each of them represents and why it has that value.

File: include/public/marco/Codegen/Lowering/Distance/SentenceDistanceCalculator.h [hunk 24]
```diff
@@ -0,0 +1,41 @@
+#ifndef MARCO_CODEGEN_LOWERING_DISTANCE_SENTENCEDISTANCECALCULATOR_H
+#define MARCO_CODEGEN_LOWERING_DISTANCE_SENTENCEDISTANCECALCULATOR_H
+
+#include "marco/Codegen/Lowering/Distance/WordDistanceCalculator.h"
+
+namespace marco::codegen::lowering
+{
+    class SentenceDistanceCalculator {
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/Distance/SentenceDistanceCalculator.h:8]: What does this class do?

File: include/public/marco/Codegen/Lowering/Distance/WordDistanceCalculator.h [hunk 25]
```diff
@@ -0,0 +1,63 @@
+#ifndef MARCO_CODEGEN_LOWERING_DISTANCE_WORDDISTANCECALCULATOR_H
+#define MARCO_CODEGEN_LOWERING_DISTANCE_WORDDISTANCECALCULATOR_H
+
+#include "marco/Codegen/Lowering/Distance/DatabaseReader.h"
+
+namespace marco::codegen::lowering
+{
+    class WordDistanceCalculator {
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/Distance/WordDistanceCalculator.h:8]: What does this class do?

File: include/public/marco/Codegen/Lowering/Distance/WordDistanceCalculator.h [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+        int combinedDistance;
+
+        // This method finds the distance to the root synset.
+        int distanceToRoot(const Synset& synset) const;
+
+    public:
+        WordDistanceCalculator();
+        ~WordDistanceCalculator();
+
+        const float alpha;
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/Distance/WordDistanceCalculator.h:29]: What are these?

File: include/public/marco/Codegen/Lowering/VariablesSymbolTable.h [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+#include <set>
+#include <string>
+
+namespace marco::codegen::lowering
+{
+  class VariablesSymbolTable {
+    public:
+      class VariablesScope {
+        public:
+          VariablesScope(VariablesSymbolTable &ht);
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/VariablesSymbolTable.h:16]: Mark as `explicit`.

File: include/public/marco/Codegen/Lowering/VariablesSymbolTable.h [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+      unsigned int count(const llvm::StringRef &Key) const;
+
+      std::optional<Reference> lookup(const llvm::StringRef &Key) const;
+
+      // If "onlyVisible" is true, return all variables ever inserted into the
+      // symbol table. Otherwise, only return current ones.
+      std::set<std::string> getVariables(bool onlyVisible) const;
+
+    private:
+      llvm::ScopedHashTable<llvm::StringRef, Reference> hashTable;
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/VariablesSymbolTable.h:33]: Rename to `references`.

File: include/public/marco/Codegen/Lowering/VariablesSymbolTable.h [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+
+      void insert(const llvm::StringRef &Key, const Reference &Val);
+
+      unsigned int count(const llvm::StringRef &Key) const;
+
+      std::optional<Reference> lookup(const llvm::StringRef &Key) const;
+
+      // If "onlyVisible" is true, return all variables ever inserted into the
+      // symbol table. Otherwise, only return current ones.
+      std::set<std::string> getVariables(bool onlyVisible) const;
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/VariablesSymbolTable.h:30]: Maybe add a default value to `onlyVisible`?

File: include/public/marco/Codegen/Lowering/VariablesSymbolTable.h [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+#include "llvm/ADT/ScopedHashTable.h"
+#include <optional>
+#include <set>
+#include <string>
+
+namespace marco::codegen::lowering
+{
+  class VariablesSymbolTable {
+    public:
+      class VariablesScope {
```
Comment (mscuttari) [inline include/public/marco/Codegen/Lowering/VariablesSymbolTable.h:14]: Where is the destructor in charge of removing the variables when exiting the scope?

**Review by mscuttari — CHANGES_REQUESTED** [review]

File: include/public/marco/Codegen/Lowering/Distance/EditDistanceParameters.h (hunk 22 above)
Comment (FrancescoPesce) [reply include/public/marco/Codegen/Lowering/Distance/EditDistanceParameters.h:34]: I will explain it more carefully in the documentation

File: include/public/marco/Codegen/Lowering/CallLowerer.h (hunk 20 above)
Comment (mscuttari) [reply include/public/marco/Codegen/Lowering/CallLowerer.h:86]: This needs to be documented.

File: include/public/marco/Codegen/Lowering/VariablesSymbolTable.h (hunk 30 above)
Comment (FrancescoPesce) [reply include/public/marco/Codegen/Lowering/VariablesSymbolTable.h:14]: If you are referring to removing the variables from `references` (formerly called `hashTable`), than, as far as I know, when `VariablesScope` is destroyed, its members' (including `scope`) destructors are called, removing the variables.
If you are instead referring to removing the variables from `variablesSet`, this is not done since it cannot be done for the reason I stated under the conversation on `ForEquationLowerer.cpp`. If case the variables need to be removed, a different solution has to be found.

Comment (mscuttari) [reply include/public/marco/Codegen/Lowering/VariablesSymbolTable.h:37]: I was referring to the latter case. However with the current design this may be acceptable.

**Review by mscuttari — CHANGES_REQUESTED** [review]
Perform the last minor changes and resolve the conflicts with the `master` branch. Then the project can be considered as completed.

**Review by mscuttari — APPROVED** [review]
